Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,628 @@
+    }
+
+    public func applicationDidEnterBackground() {
+        backgrounded = true
+    }
+
+    fileprivate func beginSyncing() {
+        notifySyncing(notification: .ProfileDidStartSyncing)
+    }
+
+    private func resolveSyncState(
+        result: MozillaAppServices.SyncResult
+    ) -> SyncDisplayState {
+        let hasSynced = !result.successful.isEmpty
+        let status = result.status
+
+        // This is similar to the old `SyncStatusResolver.resolveResults` call. If none of
+        // the engines successfully synced and a network issue occured we return `.bad`.
+        // If none of the engines successfully synced and an auth error occured we return
+        // `.warning`. Otherwise we return `.good`.
+
+        if !hasSynced && status == .authError {
+            return .warning(message: .FirefoxSyncOfflineTitle)
+        } else if !hasSynced && status == .networkError {
+            return .bad(message: .FirefoxSyncOfflineTitle)
+        } else {
+            return .good
+        }
+    }
+
+    fileprivate func endRustSyncing(_ result: MozillaAppServices.SyncResult) {
+        logger.log("Ending all syncs.",
+                   level: .info,
+                   category: .sync)
+
+        syncDisplayState = resolveSyncState(result: result)
+
+        #if MOZ_TARGET_CLIENT
+            if canSendUsageData() {
+                let gleanHelper = GleanSyncOperationHelper()
+                gleanHelper.reportTelemetry(result)
+            } else {
+                logger.log("""
+                    Profile isn't sending usage data. Not sending sync status event.
+                    """,
+                    level: .debug,
+                    category: .sync)
+            }
+        #endif
+
+        // Don't notify if we are performing a sync in the background. This prevents more
+        // db access from happening
+        if !self.backgrounded {
+            notifySyncing(notification: .ProfileDidFinishSyncing)
+        }
+    }
+
+    func canSendUsageData() -> Bool {
+        return profile.prefs.boolForKey(AppConstants.prefSendUsageData) ?? true
+    }
+
+    private func notifySyncing(notification: Notification.Name) {
+        NotificationCenter.default.post(name: notification,
+                                        object: syncDisplayState?.asObject())
+    }
+
+    init(profile: BrowserProfile,
+         logger: Logger = DefaultLogger.shared) {
+        self.profile = profile
+        self.prefs = profile.prefs
+        self.logger = logger
+
+        super.init()
+
+        let center = NotificationCenter.default
+
+        center.addObserver(self,
+                           selector: #selector(onStartSyncing),
+                           name: .ProfileDidStartSyncing,
+                           object: nil)
+        center.addObserver(self,
+                           selector: #selector(onFinishSyncing),
+                           name: .ProfileDidFinishSyncing,
+                           object: nil)
```

Comment:
Two questions related to notifications:
1. So, we send those `.ProfileDidStartSyncing` & `.ProfileDidStartSyncing` from this `RustSyncManager` class and currently from BrowserProfile code too. Will we keep sending those notifications from BrowserProfile in the near future? 
2. If we send those notifications from `RustSyncManager`, is there added benefits of listening to the notifications `RustSyncManager`? Could we directly call `onStartSyncing` and `onFinishSyncing` instead?

Reply:
1. So they're currently sent from `BrowserSyncManger` in Profile.swift. And the idea is that the `RustSyncManger` class that I'm adding in this PR will replace `BrowserSyncManager` during the Nimbus sync manager integration rollout. So there will be a lot of duplicated logic in both classes, but only one will ever be in use.
2. I suppose `RustSyncManager` doesn't need to listen to those notifications but they still need to be sent because there are a number of places in the Client code where those notifications are being observed. So, for instance, the `beginSyncing` function could call `onStartSyncing` but it would still need to emit the notification. This is more borrowed code so I don't have strong opinions here.